src/linear_algebra/matrix/reindex.lean

```diff
@@ -52,14 +57,65 @@ rfl
   reindex_linear_equiv (equiv.refl m) (equiv.refl n) = linear_equiv.refl R _ :=
 linear_equiv.ext $ λ _, rfl
 
-lemma reindex_linear_equiv_mul {o o' : Type*} [fintype o] [fintype o'] [semiring R]
+lemma reindex_linear_equiv_trans [semiring R] (e₁ : m ≃ m') (e₂ : n ≃ n') (e₁' : m' ≃ m'')
+  (e₂' : n' ≃ n'') : (reindex_linear_equiv e₁ e₂).trans (reindex_linear_equiv e₁' e₂') =
+   @reindex_linear_equiv _ _ _ _ _ _ _ _ R _ (e₁.trans e₁') (e₂.trans e₂') :=
+by { ext, refl }
+
+lemma reindex_linear_equiv_comp [semiring R] (e₁ : m ≃ m') (e₂ : n ≃ n') (e₁' : m' ≃ m'')
+  (e₂' : n' ≃ n'') : (@reindex_linear_equiv _ _ _ _ _ _ _ _ R _ e₁' e₂') ∘
+   (@reindex_linear_equiv _ _ _ _ _ _ _ _ R _ e₁ e₂)
+  = (reindex_linear_equiv (e₁.trans e₁') (e₂.trans e₂')) :=
+by { rw [← reindex_linear_equiv_trans], refl }
+
+lemma reindex_linear_equiv_comp_apply [semiring R] (e₁ : m ≃ m') (e₂ : n ≃ n') (e₁' : m' ≃ m'')
+  (e₂' : n' ≃ n'') (M : matrix m n R) :
+  (reindex_linear_equiv e₁' e₂') (reindex_linear_equiv e₁ e₂ M) =
+    reindex_linear_equiv (e₁.trans e₁') (e₂.trans e₂') M :=
+by rw [← reindex_linear_equiv_comp e₁ e₂ e₁' e₂']
+
+-- lemma reindex_linear_equiv_sum_empty_symm [semiring R] [is_empty m'] [is_empty n']
+--   (M : matrix m n R) : (reindex_linear_equiv (sum_empty m m') (sum_empty n n')).symm M =
+--     from_blocks M 0 0 0 :=
+-- begin
+--   ext (i|i) (j|j),
+--   { simp only [reindex_linear_equiv_symm, from_blocks_apply₁₁], refl },
+--   { exact is_empty_elim j },
+--   { exact is_empty_elim i }
+-- end
+
+@[simp] lemma reindex_linear_equiv_one [semiring R] [decidable_eq m] [decidable_eq m']
+  (e : m ≃ m') : (reindex_linear_equiv e e (1 : matrix m m R)) = 1 :=
+begin
```

Comment:
Can't you prove this with `reindex_one`, or does that not exist?

Reply:
I'll have a look.

Reply:
No, it was not there, but I've added it. BTW: I realise that the matrix `1 : matrix m m R` exists only assuming that `R` is a semiring; why isn't `has_one` enough? I can modify this accordingly, if it seems a good idea.

Reply:
Are there many cases where you need the identity matrix and don't have addition/multiplication of scalars? If you're willing to do the work, I'd be happy to merge a PR changing those assumptions, though I wouldn't make one myself.

By the way, don't forget to assume `has_zero` for the entries outside of the diagonal :P

Reply:
OK, you got me with the `has_zero`... I hadn't thought about it, and coming up with a "useful" example of something which has a `1`, a `0`, has an interesting matrix-theory and is not a `semiring` is unreasonable. Certainly not worth the `PR`...

Reply:
I created #7974

Reply:
>  I realise that the matrix 1 : matrix m m R exists only assuming that R is a semiring;

Citation needed. The refactor you're describing converts what you thought we had into what we have already anyway:

https://github.com/leanprover-community/mathlib/blob/1a6c871a9151444b7fd17dee134d0baffe0d5e6d/src/data/matrix/basic.lean#L173-L175